devito/petsc/solve.py

```diff
@@ -83,38 +85,77 @@ def separate_eqn(eqn, target):
     where F(target) = b.
     """
     zeroed_eqn = Eq(eqn.lhs - eqn.rhs, 0)
-    tmp = eval_time_derivatives(zeroed_eqn.lhs)
-    b, F_target = remove_target(tmp, target)
-    return -b, F_target
+    zeroed_eqn = eval_time_derivatives(zeroed_eqn.lhs)
+    target_funcs = set(generate_targets(zeroed_eqn, target))
+    b, F_target = remove_target(zeroed_eqn, target_funcs)
+    return -b, F_target, target_funcs
+
+
+def generate_targets(eq, target):
+    """
+    Extract all the functions that share the same time index as the target,
+    but may have different spatial indices.
+    """
+    funcs = retrieve_functions(eq)
+    if any(dim.is_Time for dim in target.dimensions):
+        time_idx = [
+            i for i, d in zip(target.indices, target.dimensions) if d.is_Time
+        ]
+        targets = [
+            func for func in funcs
+            if func.function is target.function and time_idx[0]
+            in func.indices
+        ]
+    else:
+        targets = [
+            func for func in funcs
+            if func.function is target.function
+        ]
+    return targets
+
+
+def targets_to_arrays(array, targets):
+    space_indices = [
+        tuple(
+            i for i, d in zip(func.indices, func.dimensions) if d.is_Space
```

Comment:
`f.space_dimensions` ?

Reply:
But I need to extract this, for example, from the list of `targets`:
`space_indices = [(x, y), (x - h_x, y), (x + h_x, y), (x, y + h_y), (x, y - h_y)]`
If there is a time index, I don't want to extract it too. Is there a cleaner way?

Reply:
But that would miss substitutions (including the use of subdomains) I think

Reply:
Why would it miss substitutions? What do you mean? Here I am just extracting the offsets, the subdomain is still picked up via the new `Eq` which is generated within `PETScSolve`.

Reply:
I think I misunderstood Fabio's comment, nvm

Reply:
Here can just do `func.indices[d] for d in func.space_dimension` as well